components/eventing-controller/pkg/subscriptionmanager/jetstream/jetstream_test.go

```diff
@@ -0,0 +1,154 @@
+		MaxReconnects:           10,
+		ReconnectWait:           time.Second,
+		EventTypePrefix:         controllertesting.EventTypePrefix,
+		JSStreamName:            controllertesting.EventTypePrefix,
+		JSStreamStorageType:     handlers.JetStreamStorageTypeMemory,
+		JSStreamRetentionPolicy: handlers.JetStreamRetentionPolicyInterest,
+	}
+	conn, err := nats.Connect(envConf.URL)
+	require.NoError(t, err)
+	jsClient, err := conn.JetStream()
+	require.NoError(t, err)
+	subsConfig := env.DefaultSubscriptionConfig{MaxInFlightMessages: 9}
+	jsBackend := handlers.NewJetStream(envConf, defaultLogger)
+	jsSubMgr.Backend = jsBackend
+
+	err = jsSubMgr.Backend.Initialize(nil)
+	require.NoError(t, err)
+
+	// create test subscription
+	testSub := controllertesting.NewSubscription("test", "test",
+		controllertesting.WithFakeSubscriptionStatus(),
+		controllertesting.WithOrderCreatedFilter(),
+		controllertesting.WithSinkURL(subscriber.SinkURL),
+		controllertesting.WithStatusConfig(subsConfig),
+	)
+
+	// create fake Dynamic clients
+	fakeClient, err := controllertesting.NewFakeSubscriptionClient(testSub)
+	require.NoError(t, err)
+	jsSubMgr.Client = fakeClient
+
+	cleaner := func(et string) (string, error) {
+		return et, nil
+	}
+	cleanedSubjects, _ := handlers.GetCleanSubjects(testSub, eventtype.CleanerFunc(cleaner))
+	testSub.Status.CleanEventTypes = jsBackend.GetJetStreamSubjects(cleanedSubjects)
+
+	// create NATS subscription
+	err = jsSubMgr.Backend.SyncSubscription(testSub)
+	require.NoError(t, err)
+
+	// make sure subscriber works
+	err = subscriber.CheckEvent("")
+	if err != nil {
+		t.Fatalf("subscriber did not receive the event: %v", err)
+	}
+	// send an event
+	err = handlers.SendEventToJetStream(jsBackend, data)
+	if err != nil {
+		t.Fatalf("publish event failed: %v", err)
+	}
+
+	// check for the event
+	err = subscriber.CheckEvent(expectedDataInStore)
+	if err != nil {
+		t.Fatalf("subscriber did not receive the event: %v", err)
+	}
+
+	// verify that the number of consumers is one
+	info, err := jsClient.StreamInfo(envConf.JSStreamName)
+	require.NoError(t, err)
+	require.Equal(t, info.State.Consumers, 1)
+
+	// when
+	err = cleanup(jsSubMgr.Backend, jsSubMgr.Client, defaultLogger.WithContext())
+
+	// then
+	require.NoError(t, err)
+
+	unstructuredSub, err := jsSubMgr.Client.Resource(controllertesting.SubscriptionGroupVersionResource()).Namespace("test").Get(ctx, testSub.Name, metav1.GetOptions{})
+	require.NoError(t, err)
+	gotSub, err := controllertesting.ToSubscription(unstructuredSub)
+	require.NoError(t, err)
+	wantSubStatus := eventingv1alpha1.SubscriptionStatus{}
+	require.Equal(t, wantSubStatus, gotSub.Status)
+
+	// test JetStream subscriptions/consumers are gone
+	info, err = jsClient.StreamInfo(envConf.JSStreamName)
+	require.NoError(t, err)
+	require.Equal(t, info.State.Consumers, 0)
+
+	// send again an event and check subscriber, check subscriber should fail
+	err = handlers.SendEventToJetStream(jsBackend, data)
```

Comment:
I wonder what the bare minimum is what the test should actually test. It feels the test is doing too much at once for me 🤔 

For me the bare minimum is that the jetstream subscription was deleted and the subscription status got updated. As done in lines 135-145.

However, testing that sending an event to a not existing jetstream subscription does not work is not necessary here I feel (it could be another test case maybe).

What is your opinion on this ?

Reply:
Sending an event to a not existing jetstream subscription is just an extension to be absolutely sure that the cleanup was done. If there are no consumers, there's no way the event can be consumed. If you also feel that this is not needed, I can remove it.

Reply:
We discussed it internally and decided it makes sense to have it, especially because the stream might get deleted later and checking the consumers is then not possible anymore.